### ISSUE TITLE
Messages not delivered and no exceptions thrown (azure-iot-device@1.4.0)

### ISSUE THREAD

**Issue by wmmihaa** [Azure/azure-iot-sdk-node#352]
# Context
- **OS and version used:** Ubuntu.Core 16.04
- **Node.js version:** 8.11
- **npm version:** 6.04

# Description of the issue:
To begin with, we have many IoT Hubs and we've only seen this issue on one of our devices on one hub. This is a new project, so we only have one device running on this particular hub. So this issue might be related to the IoT Hub and not the device SDK.

All devices are running **azure-iot-common@1.5.0,azure-iot-device@1.4.0 and azure-iot-device-mqtt@1.4.0** with **NoRetry** policy. At the particular device with this issue messages gets transmitted without any exceptions but no messages seems to get received by the IoT Hub. 

Question 1: Has you seen this behavor before? 
Question 2: As we're considering moving to 1.7, what changes has been made (I know there is a page somewhere, but I couldn't find it ;) ) 

![image](https://user-images.githubusercontent.com/3214367/44572363-e3405a80-a783-11e8-937a-225b4af9bbfc.png)

**Comment by pierreca** [comment]
@wmmihaa version 1.7.x introduces modules and edge features.
The only case that I could think of for this is:
- on linux
- the device connects successfully
- then the connection is dropped (linux doesn't let the SDK know that it happened)
- the messages keep getting enqueued and you never get a callback
- disconnect is never fired because we are never notified by the socket

we are missing a timeout logic that is separate from the retry logic, especially in the case of the mqtt library where things just get enqueued. I'm guessing the callback passed to sendEvent never gets called?

**Comment by wmmihaa** [comment]
That is the weird thing, I do get the sendEvent callback

**Comment by pierreca** [comment]
In that case it seems as though the SDK "thinks" the message was put on the wire.
You could use the azure portal to open a support ticket - the support engineers can then grab your hub name and device id, a timestamp for the issue, and look at the SLA for your hub during that time and see if anything happened on our side.

**Comment by bragma** [comment]
Hi @pierreca,
we have exactly the same problem reported above:

- on linux
- the device connects successfully
- after some time, something happens (maybe the client is disconnected)
- the messages keep getting enqueued and you never get a callback
- disconnect is never fired because we are never notified by the socket

If I try to send a direct method to the device, I get the correct reply so the device is connected / has reconnected, but still no messages are sent with sendEvent. What's the suggested fix for this? Shouldn't heartbeats solve this?

**Comment by wmmihaa** [comment]
@pierreca I missed that you had closed this issue. This is very much still an issue for us, and we have escalated it support (REF#118100419158204). Our contact is André Nascimento.

Some additional information on the subject in relation to @bragma 's comment. When we see this issue, the **Device Twin state** is "Disconnected" in the azure portal. I have not tried to call a direct method while in this state.

**Comment by anthonyvercolano** [comment]
@bragma what @wmmihaa reported is slightly different.  @wmmihaa reports that the callback for the send is invoked.  I am assuming that the callback is invoked with no error.

@bragma the information that you can invoke a device method is interesting in that it certainly implies a live connection to the device.

@bragma you would think that lack of PINGRESP would cause a disconnect.  Maybe (I don't know) the PINGREQ/PINGRESP was still happening?

**Comment by bragma** [comment]
Ok I confirm my situation is different. At some time, sendEvent stops calling the callback and never resumes. The device is connected, I can send direct messages to it and get the answers.
I've had to explicitly set policy to NoRetry an then after 6-7 minutes, I finally get an error but only for last sendEvent call. All previous calls and their callback are eaten up somewhere and I fear of some long term resource/memory leak.

**Comment by wmmihaa** [comment]
Good morning (or night),

We have currently two devices that are offline in the Azure portal but seams to think they are online and keep sending data which is not coming through. Do you which to investigate this now, in which case I can give you access. Please respond as soon as possible as I otherwise need to restart the devices.

Mikael

**Comment by anthonyvercolano** [comment]
Generally the repo/issues is only monitored by the sdk developers in Redmond.  Not likely to see anything between 10pm and 7am us west coast time.

@wmmihaa I would suggest you restart the devices.  Can you send me email with the hub name and devices?  Also as precise a time as is possible when the sends stopped showing up in the hub?

**Comment by anthonyvercolano** [comment]
Sorry forgot that it looks like you have an ICM also, perhaps you should send the info to Andre so that this goes in the ICM.

**Comment by anthonyvercolano** [comment]
Also, as an aside, have you tried using HTTP as the protocol?  If so, did that ever exhibit this problem?

**Comment by wmmihaa** [comment]
@anthonyvercolano André is fully updated on all findings

**Comment by wmmihaa** [comment]
@anthonyvercolano no I have not tried HTTP (only MQTT, MQTT-WS, AMQP, AMQP-WS). Worth notice ... we are using both state and direct methods.

**Comment by anthonyvercolano** [comment]
One thing that helps is if we can get internal logs:

Setting the environment variable DEBUG to "amqp*,azure*"

for the version of the SDK code that you're running now will provide a great deal of information.

However, one thing I would also suggest is to try updating to the newer SDK.  We believe we've made some nice improvements and are also using a different AMQP implmenentation.  A significant amount of work went into handling disconnects better.  If you did update to the latest (a new one is likely today), it would be necessary to set the DEBUG environment variable to "rhea*,azure*".

**Comment by wmmihaa** [comment]
Thanks @anthonyvercolano for great feedback. 
>One thing that helps is if we can get internal logs:
Setting the environment variable DEBUG to "amqp*,azure*"

The problem is that this behavior only happens once or twice among all our devices, and I would therefor have to enable DEBUG on ALL of them. Not only do we have many devices, we also run many IoT Hubs (all experiancing the same issue.
**Du you know if there is a way to set this flag while the process is running?** 

> However, one thing I would also suggest is to try updating to the newer SDK.

I raised this question earlier in this thread, but understood there had not been any changes in regards to this matter, hence we haven't considered making the update. Please have a look at @pierreca 's respons (second from the top) 
**If you can confirm that that is the case, I'm happy to make the update.**. Also, if this is the case, do you know if you've resolved the "*disconnect is never fired because we are never notified by the socket*" (MQTT on linux).

**Comment by anthonyvercolano** [comment]
Linux socket disconnects are particularly difficult to deal with wrt to disconnects.  Different philosophy then windows.  Still both AMQP and MQTT keep alive should be reacting to unanswered pings.  This will be a significant area of investigation in the coming days.

At the beginning of this thread this it is reported to be using MQTT.  The most recent configuration appears to be doing AMQP per your note of this morning.  We didn't switch over to the new transport till September.

Nothing has really changed in the MQTT transport.  Sadly that transport doesn't include any debug logs.

I don't know of anyway to set the environment variable DEBUG and have an already running node application notice the new DEBUG value.  I've tried.

The logs would be helpful.  Whatever you can do would be appreciated.

**Comment by wmmihaa** [comment]
> At the beginning of this thread this it is reported to be using MQTT.

Yes, we were initially running MQTT, but changed to AMQP after being recommended doing so from the product team. However, we're still suffering from what appears to be the same issue but not as often as with MQTT.

>However, one thing I would also suggest is to try updating to the newer SDK. We believe we've made some nice improvements and are also using a different AMQP implementation. A significant amount of work went into handling disconnects better.

Looking into the repo, I don't see much activity since the issue was created, but perhaps these changes were made in underlying C/C++ libraries. But would you mind confirming this, and I'll immediately update all devices to use the latest stable version.

Again, thanks for all the help.

**Comment by bragma** [comment]
@anthonyvercolano first of all thank you for your efforts. Connection stability is really important, also knowing the correct procedure to reconnect without causing memory/resource leaks is important. As I said elsewere, the problem with examples is that most of them are just "happy paths" without care for production environments. As more and more companies move to IoT solutions, we really need software that can cope with problems and run for months without issues. At the moment I have real doubts the current libraries have these requisites (and the same can be said for C# versions of the SDK).

My setup is very simple: I use a sample node client created with VSCode and IoT Hub package. It is fairly basic stuff, just sending D2C telemetry. At the moment I do a very simple thing: I remove the network cable (but also using a firewall will do) and I need to detect that it happened, the sooner the better. It's ok if heartbeats are required and it takes some time. But when it happens, I also need a proper way to recover, retry and handle the situation.

At the moment, with the default retry policy, disconnecting the cable nothing happens. My app thinks it's still sending telemetry happily while in fact it is not. I've had some better results changing the policy to NoRetry. It still takes time, but eventually I receive some errors. The problem there is that the callbacks associated to "lost" sendEvent calls are never called back. They are lost somewhere. At this point, I can handle the disconnection but I have 2 problems:

- Allocated callbacks will probably cause leaks on the long run
- I don't know what's been send and what not

Notice that I can't switch to http transpost, I need a somewhat real time 2 way connection for direct methods with the device.

I think the entire node clients really needs some reenginereing with a strong attention to handling those problems. IoT Hub is a good product as much as its clients are.

Thanks a lot!

**Comment by anthonyvercolano** [comment]
@wmmihaa I believe that the current AMQP base client is an improvement.  Spent a significant amount of time in disconnect testing when changing over to the rhea amqp library.  In addition, the debugging available is quite nice from rhia.

@bragma I could not agree more that the heart and soul of the SDK should be the care and maintenance of the connection.  It's hard and quite frankly a significant value add for developers.

@bragma I don't recall which version of the SDK you're currently running.  At least in the case of AMQP there is new code in the SDK to run down all pending D2C messages that are currently on the wire (or more specifically, un-disposed).  If using the current SDK and these don't get run down in the case of an actual disconnect, I would certainly be chagrined .

**Comment by wmmihaa** [comment]
> I believe that the current AMQP base client is an improvement. Spent a significant amount of time in disconnect testing when changing over to the rhea amqp library. In addition, the debugging available is quite nice from rhea.

Thank you for all the help. I'll keep you updated, and will close the issue.

**Comment by az-iot-builder-01** [comment]
@bragma, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by bragma** [comment]
> @bragma I don't recall which version of the SDK you're currently running. At least in the case of AMQP there is new code in the SDK to run down all pending D2C messages that are currently on the wire (or more specifically, un-disposed). If using the current SDK and these don't get run down in the case of an actual disconnect, I would certainly be chagrined .

I am using MQTT because I have evaluated pros and cons with AMQP and decided to go with the former.

    "azure-iot-device-mqtt": "^1.8.1",
    "azure-iot-device": "^1.8.1"

**Comment by anthonyvercolano** [comment]
Thank you for the logs.  Hefty.   Please send me the hub name in **email**.  With the hub name I can look up in service logs for events that might have occurred.  I have my suspicions.

The only way the repeating debug statement could happen is if the underlying amqp transport thinks that it can't send anymore data.  (Line 365 of common/transport/amqp/src/sender_link.ts)

The twin IMPORTANT information.  This is what you see at the same time that the sends aren't happening?  This is actually a good clue too.  Twins "live" in a different place in the service.  I can imagine scenarios where twins continue to work and the sends "stop" and also the other way, where twins stop working and sends continue.

Don't upgrade anymore gateways.

Change the debug variable setting to "azure*,rhea*"

Expectations:

Basically everyone on the team will be starting vacations today or tomorrow.  I will be out of touch in Asia until the 7th.

**Comment by wmmihaa** [comment]
anthony.v.ercolano@smallAndSoft.com?

I'll update the Debug info, and I'll see if I can find some more debug data in the sender_link.js

**Comment by wmmihaa** [comment]
FYI... I don't see any "rhea" debug output...

Currently...all I see is:
```2018-12-17T15:33:57.268Z azure-iot-amqp-base:SenderLink in sender attached state - ...
2018-12-17T15:33:58.097Z azure-iot-device-amqp:Amqp using existing d2c link                            
2018-12-17T15:33:58.099Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue. 
2018-12-17T15:33:58.100Z azure-iot-amqp-base:SenderLink unsent message queue length is: 1              
2018-12-17T15:33:58.100Z azure-iot-amqp-base:SenderLink message placed in dictionary for lookup later. ```

**Comment by anthonyvercolano** [comment]
/microservicebus-node/32/node_modules/azure-iot-device-amqp/node_modules/azure-iot-amqp-base/lib/sender_link.js  ?

**Comment by anthonyvercolano** [comment]
As to the lack of rhea output?  Restart?

**Comment by wmmihaa** [comment]
[New log file](https://microservicebusstorage.blob.core.windows.net/syslogs/jh/syslog2.tar.gz)

![image](https://user-images.githubusercontent.com/3214367/50097842-07381800-021b-11e9-8036-f0f5fbc86f82.png)

**Comment by anthonyvercolano** [comment]
Device ID? (In email)

**Comment by wmmihaa** [comment]
> As to the lack of rhea output? Restart?

Already did... :(

**Comment by wmmihaa** [comment]
I've added the following line to the /rhea/lib/session.js:46:
``` javascript
CircularBuffer.prototype.available = function () {
    console.log("AZURE DEBUGINFO: capacity: " + this.capacity + " size:" + this.size + ' head=' + this.head + ' tail=' + this.tail + " available:" + (this.capacity - this.size));
    return this.capacity - this.size;
};
```

hopefully this should give you some good input...

**Comment by wmmihaa** [comment]
Ok... this looks interesting:
mSB: DEBUG: [azureIoTEventHub1]=>Event submitted to Azure IoT Hub                                                
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:14.636Z azure-iot-amqp-base:SenderLink in sender attached state - accepted event for rheaSender_
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:15.019Z azure-iot-device-amqp:Amqp using existing d2c link                                      
2018-12-17T18:31:15.019Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue.           
AZURE DEBUGINFO: capacity: 2048 size:2044 head=1 tail=2045 available:4                                           
2018-12-17T18:31:15.021Z azure-iot-amqp-base:SenderLink unsent message queue length is: 1                        
2018-12-17T18:31:15.022Z azure-iot-amqp-base:SenderLink message placed in dictionary for lookup later.           
mSB: DEBUG: [azureIoTEventHub1]=>Event submitted to Azure IoT Hub                                                
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:15.432Z azure-iot-amqp-base:SenderLink in sender attached state - accepted event for rheaSender_
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:15.520Z azure-iot-device-amqp:Amqp using existing d2c link                                      
2018-12-17T18:31:15.521Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue.           
AZURE DEBUGINFO: capacity: 2048 size:2045 head=1 tail=2046 available:3                                           
2018-12-17T18:31:15.522Z azure-iot-amqp-base:SenderLink unsent message queue length is: 1                        
2018-12-17T18:31:15.523Z azure-iot-amqp-base:SenderLink message placed in dictionary for lookup later.           
mSB: DEBUG: [azureIoTEventHub1]=>Event submitted to Azure IoT Hub                                                
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:15.641Z azure-iot-amqp-base:SenderLink in sender attached state - accepted event for rheaSender_
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:16.026Z azure-iot-device-amqp:Amqp using existing d2c link                                      
2018-12-17T18:31:16.026Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue.           
AZURE DEBUGINFO: capacity: 2048 size:2046 head=1 tail=2047 available:2                                           
2018-12-17T18:31:16.027Z azure-iot-amqp-base:SenderLink unsent message queue length is: 1                        
2018-12-17T18:31:16.028Z azure-iot-amqp-base:SenderLink message placed in dictionary for lookup later.           
mSB: DEBUG: [azureIoTEventHub1]=>Event submitted to Azure IoT Hub                                                
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:16.143Z azure-iot-amqp-base:SenderLink in sender attached state - accepted event for rheaSender_
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:16.528Z azure-iot-device-amqp:Amqp using existing d2c link                                      
2018-12-17T18:31:16.531Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue.           
AZURE DEBUGINFO: capacity: 2048 size:2047 head=1 tail=0 available:1                                              
2018-12-17T18:31:16.533Z azure-iot-amqp-base:SenderLink unsent message queue length is: 1                        
2018-12-17T18:31:16.535Z azure-iot-amqp-base:SenderLink message placed in dictionary for lookup later.           
mSB: DEBUG: [azureIoTEventHub1]=>Event submitted to Azure IoT Hub                                                
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:16.647Z azure-iot-amqp-base:SenderLink in sender attached state - accepted event for rheaSender_
AZURE DEBUGINFO: capacity: 2048 size:5 head=2 tail=7 available:2043                                              
2018-12-17T18:31:17.035Z azure-iot-device-amqp:Amqp using existing d2c link                                      
2018-12-17T18:31:17.035Z azure-iot-amqp-base:SenderLink placing a message in the unsent message queue.           
AZURE DEBUGINFO: capacity: 2048 size:2048 head=1 tail=1 available:0                                              
``````
Please notice that the **CircularBuffer.prototype.available** is called three times per **deviceClient.sendEvent**. One of the three is constantly decreasing until the point it's "0" and becomes unavailable. Hence stop transmitting messages. 5-10 later the device twin becomes "Disconnected".

I started the device at 06:00 and the error happened at 18:31:17. 

[Download log](https://microservicebusstorage.blob.core.windows.net/syslogs/jh/syslog3.tar.gz)

**Comment by wmmihaa** [comment]
Here is a working sample for you. Might simplify debugging 😉

Note that I got this issue when I add the twin... *deviceClient.getTwin*:22 (app.js). Also...I've added the same console output to the *CircularBuffer.prototype.available* call

https://microservicebusstorage.blob.core.windows.net/syslogs/jh/sample.zip

**Comment by anthonyvercolano** [comment]
You're doing quite a bit of good work.

So this example eventually stops sending?  You only ask for desired property updates. Do you actually have the twin properties updating?

**Comment by wmmihaa** [comment]
2048 - intervalOfSendingEvents ;)

The Session::size will increase for every call, which decreases the availablity(available === capacity - size). 
So you can send 2048 messages until it fails...

**Comment by anthonyvercolano** [comment]
I've opened up an issue in the rhea repo:

[rhea isssue 171](https://github.com/amqp/rhea/issues/171)

**Comment by anthonyvercolano** [comment]
It may very well be that I've forgotten to do something to "free" up that send.

**Comment by wmmihaa** [comment]
Although I still think it might be related to your package, as it cleearly relates to twin. If you remove the getTwin... it all works

**Comment by anthonyvercolano** [comment]
Wait, you told me that I would get this if I didn't do the twin?

**Comment by wmmihaa** [comment]
> Note that I got this issue when I add the twin... deviceClient.getTwin:22 (app.js). Also...I've added the same console output to the CircularBuffer.prototype.available call

> Actually the you can disregard the listening for desired properties 
Meaning you can disregard :
```
 twin.on('properties.desired', function (desiredChange) {})
```

**Comment by anthonyvercolano** [comment]
I did leave the twin call in the sample you sent.  I have indeed hit the problem.

**Comment by wmmihaa** [comment]
This is going to be the longest thread in history of GitHub... you should use gitter ;)

**Comment by anthonyvercolano** [comment]
You are saying reasonable things.  I'm just not yet seeing how the twin link creation/use pollutes the sender link for the d2c.  In your sample all the twin operations happen VERY early.  Then nothing from twin from then on.  However, in each case of a send after that the available count does go down.  Yet we absolutely see something from the server that marks the send message as settled, yet the available count doesn't go back up!

If I indeed remove the twin code from the sample, the available count stays a rock steady 2048.

**Comment by wmmihaa** [comment]
Ok... I'll keep digging. Let me know if there is anything you'd like me to focus on.
